starky2/src/cross_table_lookup.rs

```diff
@@ -0,0 +1,299 @@
+                    beta,
+                    gamma,
+                );
+                let z_looked = partial_products(
+                    &trace_poly_values[*looked_table as usize],
+                    looked_columns,
+                    beta,
+                    gamma,
+                );
+
+                debug_assert_eq!(
+                    *z_looking.values.last().unwrap(),
+                    *z_looked.values.last().unwrap()
+                        * (gamma + reduce_with_powers(default.iter(), beta)).exp_u64(
+                            trace_poly_values[*looking_table as usize][0].len() as u64
+                                - trace_poly_values[*looked_table as usize][0].len() as u64
+                        )
+                );
+
+                acc[*looking_table as usize]
+                    .zs_columns
+                    .push((z_looking, looking_columns.clone()));
+                acc[*looked_table as usize]
+                    .zs_columns
+                    .push((z_looked, looked_columns.clone()));
+            }
+            acc
+        },
+    )
+}
+
+fn partial_products<F: Field>(
+    trace: &[PolynomialValues<F>],
+    columns: &[usize],
+    beta: F,
+    gamma: F,
+) -> PolynomialValues<F> {
+    let mut partial_prod = F::ONE;
+    let degree = trace[0].len();
+    let mut res = Vec::with_capacity(degree);
+    for i in 0..degree {
+        partial_prod *=
+            gamma + reduce_with_powers(columns.iter().map(|&j| &trace[j].values[i]), beta);
+        res.push(partial_prod);
+    }
+    res.into()
+}
+
+#[derive(Clone)]
+pub struct CtlCheckVars<'a, F, FE, P, const D2: usize>
+where
+    F: Field,
+    FE: FieldExtension<D2, BaseField = F>,
+    P: PackedField<Scalar = FE>,
+{
+    pub(crate) local_z: P,
+    pub(crate) next_z: P,
+    pub(crate) challenges: GrandProductChallenge<F>,
+    pub(crate) columns: &'a [usize],
+}
+
+impl<'a, F: RichField + Extendable<D>, const D: usize>
+    CtlCheckVars<'a, F, F::Extension, F::Extension, D>
+{
+    pub(crate) fn from_proofs<C: GenericConfig<D, F = F>>(
+        proofs: &[&StarkProofWithPublicInputs<F, C, D>],
+        cross_table_lookups: &'a [CrossTableLookup<F>],
+        ctl_challenges: &'a GrandProductChallengeSet<F>,
+        num_permutation_zs: &[usize],
+    ) -> Vec<Vec<Self>> {
+        debug_assert_eq!(proofs.len(), num_permutation_zs.len());
+        let mut ctl_zs = proofs
+            .iter()
+            .zip(num_permutation_zs)
+            .map(|(p, &num_permutation)| {
+                p.proof
+                    .openings
+                    .permutation_ctl_zs
+                    .iter()
+                    .skip(num_permutation)
+                    .zip(
+                        p.proof
+                            .openings
+                            .permutation_ctl_zs_right
+                            .iter()
+                            .skip(num_permutation),
+                    )
+            })
```

Comment:
nit: I think it'd be easier to read with a few intermediate vars -

```suggestion
            .map(|(p, &num_perms)| {
                let openings = &p.proof.openings;
                let ctl_zs = openings.permutation_ctl_zs.iter().skip(num_perms);
                let ctl_zs_right = openings.permutation_ctl_zs_right.iter().skip(num_perms);
                ctl_zs.zip(ctl_zs_right)
            })
```